File: packages/document-methods/src/llm_conversation.ts
import { v7 } from "uuid";

import type {
    EvalResult,
    FeedbackResolution,
    InlineFile,
    Document,
    LLMInteraction,
    StableRef,
    Uuid,
} from "catcolab-document-types";
import { currentVersion } from "catcolab-document-types";

/** A document containing an LLM conversation attached to a model. */
export type LLMConversationDocument = Document & { type: "llmconversation" };

/** Create an empty LLM conversation for a document. */
export const newLLMConversationDocument = (
    documentRef: StableRef,
    llmModel: string,
): LLMConversationDocument => ({
    name: "",
    type: "llmconversation",
    llmConversationOf: {
        ...documentRef,
        type: "llmconversation-of",
    },
    llmModel,
    interactions: [],
    version: currentVersion(),
});

/** Create an interaction containing a user message. */
export function newUserMessage(
    content: string,
    files: InlineFile[],
): Extract<LLMInteraction, { tag: "user-message" }> {
    return {
        tag: "user-message",
        id: v7(),
        timestamp: new Date().toISOString(),
        content,
        files,
    };
}

/** Create an interaction containing a completed LLM message. */
export function newLLMMessage(content: string): Extract<LLMInteraction, { tag: "llm-message" }> {
    return {
        tag: "llm-message",
        id: v7(),
        timestamp: new Date().toISOString(),
        content,
    };
}

/** Create an interaction containing one completed `contextExec` call. */
export function newLLMCodeExecution(
    toolCallId: string,
    code: string,
    result: EvalResult,
    transaction?: unknown,
): Extract<LLMInteraction, { tag: "llm-code-execution" }> {
    return {
        tag: "llm-code-execution",
        id: v7(),
        timestamp: new Date().toISOString(),
        toolCallId,
        code,
        result,
        ...(transaction === undefined ? {} : { transaction }),
    };
}

/** Append an interaction to a conversation's ordered history. */
export function appendLLMInteraction(
    conversation: LLMConversationDocument,
    interaction: LLMInteraction,
) {
    conversation.interactions.push(interaction);
}

/** Reject every pending feedback request superseded by a new user message. */
export function rejectUnresolvedUserFeedbackRequests(
    conversation: LLMConversationDocument,
): number {
    let rejected = 0;
    for (const interaction of conversation.interactions) {
        if (
            interaction.tag === "user-feedback-request" &&
            interaction.resolution === "unresolved"
        ) {
            interaction.resolution = "rejected";
            rejected += 1;
        }
    }
    return rejected;
}

/** Resolve a pending feedback request by its stable interaction ID. */
export function resolveUserFeedbackRequest(
    conversation: LLMConversationDocument,
    requestId: Uuid,
    resolution: Exclude<FeedbackResolution, "unresolved">,
) {
    const request = conversation.interactions.find(
        (interaction): interaction is Extract<LLMInteraction, { tag: "user-feedback-request" }> =>
            interaction.tag === "user-feedback-request" && interaction.id === requestId,
    );
    if (!request || request.resolution !== "unresolved") {
        return false;
    }
    request.resolution = resolution;
    return true;
}

File: packages/document-types/src/v2/llm_conversation.rs
use super::api::Link;

use serde::{Deserialize, Serialize};
use serde_json::Value;
use tsify::Tsify;
use uuid::Uuid;

/// A supported inline file type.
#[derive(PartialEq, Eq, Debug, Serialize, Deserialize, Tsify)]
#[tsify(into_wasm_abi, from_wasm_abi)]
pub enum FileType {
    CSV,
}

/// A file stored inline with a user message.
#[derive(PartialEq, Eq, Debug, Serialize, Deserialize, Tsify)]
#[tsify(into_wasm_abi, from_wasm_abi)]
pub struct InlineFile {
    pub filename: String,
    #[serde(rename = "fileType")]
    pub file_type: FileType,
    pub content: Vec<u8>,
}

/// The result of executing code requested by the LLM.
#[derive(PartialEq, Eq, Debug, Serialize, Deserialize, Tsify)]
#[serde(tag = "tag")]
#[tsify(into_wasm_abi, from_wasm_abi)]
pub enum EvalResult {
    Ok { value: String },
    Err { error: String },
}

/// A message submitted by the user.
#[derive(PartialEq, Eq, Debug, Serialize, Deserialize, Tsify)]
#[tsify(into_wasm_abi, from_wasm_abi)]
pub struct UserMessage {
    pub timestamp: String,
    pub id: Uuid,
    pub content: String,
    pub files: Vec<InlineFile>,
}

/// A completed textual response from the LLM.
#[derive(PartialEq, Eq, Debug, Serialize, Deserialize, Tsify)]
#[tsify(into_wasm_abi, from_wasm_abi)]
pub struct LLMMessage {
    pub timestamp: String,
    pub id: Uuid,
    pub content: String,
}

/// A `contextExec` call and its completed result.
#[derive(PartialEq, Eq, Debug, Serialize, Deserialize, Tsify)]
#[tsify(into_wasm_abi, from_wasm_abi)]
pub struct LLMCodeExecution {
    pub timestamp: String,
    pub id: Uuid,
    #[serde(rename = "toolCallId")]
    pub tool_call_id: String,
    pub code: String,
    pub result: EvalResult,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub transaction: Option<Value>,
}

/// The user's resolution of a feedback request.
#[derive(PartialEq, Eq, Debug, Serialize, Deserialize, Tsify)]
#[serde(rename_all = "lowercase")]
#[tsify(into_wasm_abi, from_wasm_abi)]
pub enum FeedbackResolution {
    Unresolved,
    Approved,
    Rejected,
}

/// A request for the user to approve or reject a proposed transaction.
#[derive(PartialEq, Eq, Debug, Serialize, Deserialize, Tsify)]
#[tsify(into_wasm_abi, from_wasm_abi)]
pub struct UserFeedbackRequest {
    pub timestamp: String,
    pub id: Uuid,
    #[serde(rename = "codeExecution")]
    pub code_execution: Uuid,
    pub content: String,
    pub resolution: FeedbackResolution,
}

/// An interaction in an LLM conversation.
#[derive(PartialEq, Eq, Debug, Serialize, Deserialize, Tsify)]
#[serde(tag = "tag")]
#[tsify(into_wasm_abi, from_wasm_abi)]
pub enum LLMInteraction {
    #[serde(rename = "user-message")]
    UserMessage(UserMessage),
    #[serde(rename = "llm-message")]
    LLMMessage(LLMMessage),
    #[serde(rename = "llm-code-execution")]
    LLMCodeExecution(LLMCodeExecution),
    #[serde(rename = "user-feedback-request")]
    UserFeedbackRequest(UserFeedbackRequest),
}

/// A sequential conversation attached to a CatColab model.
#[derive(PartialEq, Eq, Debug, Serialize, Deserialize, Tsify)]
#[tsify(into_wasm_abi, from_wasm_abi)]
pub struct LLMConversationDocumentContent {
    pub name: String,
    #[serde(rename = "llmConversationOf")]
    pub conversation_of: Link,
    #[serde(rename = "llmModel")]
    pub llm_model: String,
    pub interactions: Vec<LLMInteraction>,
    pub version: String,
}
